test/corpus/literals.txt

```diff
@@ -454,6 +432,131 @@ cat ${BAR} ${ABC=def} ${GHI:?jkl}
           (variable_name)
           (regex))))))
 
+================================================================================
+Variable Expansions: Length
+================================================================================
+
+${parameter:-1}
+
+${parameter: -1}
+
+${parameter:(-1)}
+
+${matrix:$(($random%${#matrix})):1}
+
+"${_component_to_single:${len}:2}"
+
+--------------------------------------------------------------------------------
+
+(program
+  (command (command_name (expansion (variable_name) (word))))
+  (command (command_name (expansion (variable_name) (number))))
+  (command
+    (command_name
+      (expansion (variable_name) (parenthesized_expression (unary_expression (number))))))
+  (command
+    (command_name
+      (expansion
+        (variable_name)
+        (arithmetic_expansion
+          (binary_expression (simple_expansion (variable_name)) (expansion (variable_name))))
+        (number))))
+  (command (command_name (string (expansion (variable_name) (expansion (variable_name)) (number))))))
+
+================================================================================
+Variable Expansions with operators
+================================================================================
+
+${parameter-default}
+${parameter- default}
+${!varprefix*}
+${!varprefix@}
+${parameter@U}
+
+--------------------------------------------------------------------------------
+
+(program
+  (command (command_name (expansion (variable_name) (word))))
+  (command (command_name (expansion (variable_name) (word))))
+  (command (command_name (expansion (variable_name))))
+  (command (command_name (expansion (variable_name))))
+  (command (command_name (expansion (variable_name)))))
+
+================================================================================
+Variable Expansions: Bizarre Cases
+================================================================================
+
+${!#}
+${!# }
+${!##}
+${!## }
+${!##/}
+
+--------------------------------------------------------------------------------
+
+(program
+  (command (command_name (expansion)))
+  (command (command_name (expansion)))
+  (command (command_name (expansion)))
+  (command (command_name (expansion)))
+  (command (command_name (expansion (special_variable_name) (regex)))))
+
+================================================================================
+Variable Expansions: Weird Cases
+================================================================================
+
+${completions[*]}
+${=1}
```

Comment:
`bash: ${=1}:: bad substitution`

Reply:
It's valid in here, let me find the example file that had it

```bash
for c in ${=1}; do
	echo c
done
```

Reply:
http://www.oilshell.org/git-branch/master/21552818/wild.wwz/linux-4.8.7/tools/perf/perf-completion.sh.txt

Reply:
Interesting. I can't find it anywhere in the bash manual. :thinking:

Reply:
maybe it's in an older version